README.md
# Rspec::MultiprocessRunner

This gem provides a mechanism for running a suite of RSpec tests in multiple
processes on the same machine and multiple machines (all reporting back to the
head node [machine]), potentially allowing substantial performance improvements.

It differs from `parallel_tests` in that it uses a coordinator process on each
machine to manage the workers, hand off work to them, and receive results. These
coordinators communicate via simple TCP messages with a head node coordinator to
remain in sync when running on multiple machines. This means it can dynamically
balance the workload among the processors and machines. It also means it can
provide consolidated results in one console.

It follows parallel_tests' environment variable conventions so it's easy to
use them together. (E.g., parallel_tests has a very nice set of rake tasks
for setting up parallel environments.)

## Benefits

* Running slow (IO-bound) specs in parallel can greatly reduce the wall-clock
  time needed to run a suite. Even CPU-bound specs can be aided
* Provides detailed logging of each example as it completes, including the
  failure message (you don't have to wait until the end to see the failure
  reason). This is only on the machine running the spec, a final print out does
  occur on the main machine upon completion, however.
* Detects, kills, and reports spec files that take longer than expected (five
  minutes by default).
* Detects and reports spec files that crash (without interrupting the
  remainder of the suite).
* Head node detects and reruns spec files that are not reported back by the
  nodes.

## Limitations

* Only works with RSpec 3 (>= 2.99). Does not work with previous versions of
  RSpec.
* Does not work on Windows or JRuby. Since it relies on `fork(2)`, it probably
  never will.
* Does not support RSpec custom formatters.
* The built-in output format is very verbose — it's intended for CI servers,
  where more logging is better.
* Intermediate-quality code. Happy path works, and workers are
  managed/restarted, but:
  * There's no test coverage of the runner itself, only auxiliaries.
* No security in the TCP messaging, but can work over SSH tunnels. Nodes do
  verify the file name given against the known files so it will not execute
  maliciously.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'rspec-multiprocess_runner'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rspec-multiprocess_runner

## Usage

### Command line

Use `multirspec` to run a bunch of spec files:

    $ multirspec spec

Runs three workers by default — use `-w` to chose another count. `--help` will
detail the other options.

You can provide options that will be passed to the separate RSpec processes by
including them after a `--`:

    $ multirspec -w 5 spec -- -b -I ./lib

In this case, each RSpec process would receive the options `-b -I ./lib`. Note
that not that many RSpec options really make sense to pass this way. In
particular, file selection and output formatting options are unlikely to work
the way you expect.

Runs as a head node node by default. The following command mimics the defaults:

    $ multirspec -p 2222 -n 5 [Files]

A corresponding node node, for a head node on `head_node.local` would be:

    $ multirspec -H head_node.local -p 2222 -n [Files]

N.B. You must include the same files for the nodes as the head node.

A corresponding set up for a node node using SSH would be:

    $ ssh -nNT -L 2500:localhost:2222 head_node.local &
    $ multirspec -H localhost -p 2500 -n [Files]
    $ kill $(jobs -p)

N.B. Ensure that the head node has tcp local port forwarding permitted.

A set of nodes and head node can be given a unique string in case multiple spec
file versions may be tested near simultaneously to prevent execution of the
wrong version.

    $ multirspec -r a_git_commit_hash

### Rake

There is a rake task wrapper for `multirspec`:

    require 'rspec/multiprocess_runner/rake_task'

    RSpec::MultiprocessRunner::RakeTask.new(:spec) do |t|
      t.worker_count = 5
      t.pattern = "spec/**/*_spec.rb"
    end

See its source for the full list of options.

### Code

Create a coordinator and tell it to run:

    require 'rspec/multiprocess_runner/coordinator'

    worker_count = 4
    per_file_timeout = 5 * 60    # 5 minutes in seconds
    rspec_options = %w(--backtrace)
    files = Dir['**/*_spec.rb']

    coordinator = RSpec::MultiprocessRunner::Coordinator.new(
      worker_count, files,
      {
        file_timeout_seconds: per_file_timeout,
        rspec_options: rspec_options
      }
    )
    coordinator.run

… but you are probably better off using the command line interface.

## Runtime environment

### `TEST_ENV_NUMBER`

This runner provides each worker with a environment number so that you can
segment access to resources (databases, etc). For convenience, it follows the
same convention as `parallel_tests`.

* The environment number is provided as an environment variable named `TEST_ENV_NUMBER`
* The first environment number is `""` and subsequent ones are integers (`"2"`, `"3"`, etc.)

Like `parallel_tests` 2.3 and later, `multirspec` supports a `--first-is-1`
argument which makes the first environment number `"1"` instead. Use this to
have your multiprocess runs be isolated from the environment used in tests you
run normally, allowing you to to TDD while a long-running multiprocess run
continues in the background.

### Behavior options

All options are available via the `multirspec` command line interface. A couple
may alternatively be set in the calling environment.

* Worker count: instead of providing `-w 5`, set `PARALLEL_TEST_PROCESSORS="5"` or
  `MULTIRSPEC_WORKER_COUNT="5"` in the environment. If both `-w` and one of these
  vars is set, the value passed to `-w` will win.
* First-is-1: instead of providing `--first-is-1`, set
  `PARALLEL_TEST_FIRST_IS_1="true"` or `MULTIRSPEC_FIRST_IS_1="true"` in the
  environment.

These environment variables only apply to the CLI and rake task. If you
directly invoke `RSpec::MultiprocessRunner::Coordinator#run`, they are ignored.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/cdd/rspec-multiprocess_runner.

### Project infrastructure

* [Source on GitHub](https://github.com/cdd/rspec-multiprocess_runner)
* [![Build Status](https://travis-ci.org/cdd/rspec-multiprocess_runner.svg?branch=master)](https://travis-ci.org/cdd/rspec-multiprocess_runner)
  [Continuous Integration on Travis-CI](https://travis-ci.org/cdd/rspec-multiprocess_runner)

### Release process

1. Verify that all desired changes have been merged & pushed to master.
2. Verify that the changelog is up to date (it should be kept up to date as
   changes are made, so this should just be a quick check).
3. Verify that the current master has passed on Travis.
4. Edit `version.rb` and remove `".pre"` from the version number. Save and commit.
5. Run `rake release`. This packages the gem and submits it to rubygems.org.
6. Edit `version.rb` and update to the next patch-level release, plus `.pre`.
   E.g. if you just released 0.6.10, update the version in `version.rb` to
   `"0.6.11.pre"`.
7. Add a heading for the new version number to `CHANGELOG.md`. E.g., if you
   just released 0.6.10, add "# 0.6.11" to the top of the changelog.
8. Save and commit the changes from steps 6 and 7.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# 1.4.4

# 1.4.3

* New 1.4.2 logging is less aggressive and should only show errors now
* Adds a docker container for releasing

# 1.4.2

* Add better logging for errors that occur outside of tests
* Add docker tests for all current ruby versions and more ruby versions to test in
* Fix some minor bugs that caused incorrect error codes (still failure codes)
* Fixed the error code spec file

# 1.4.1

* Fix file coordinator helpers failing due to unknown method on command line args

# 1.4.0

* Fix issue catching spec files failing outside of tests (it will now die loudly instead of passing silently).

# 1.3.3

* File output bug fix

# 1.3.1

* Or use an environment variable for summary file.

# 1.3.0

* Add file output option

# 1.2.3

* Fix time library sporadic failure

# 1.2.2

* Fix typo in rake task

* Also adds specs that should have been included in last revision

# 1.2.1

* Fix missing files being triggered by empty files

* Fix missing files also being skipped files

* Add run identifier option to prevent incorrect spec version execution

# 1.2.0

* Can now run tests on multiple machines at a time by running in head node mode
  and others in node mode.  Nodes get which files to run when from the head node
  in a manner similar to how workers got files from the coordinator (which
  currently also still happens).

* Uses unencoded TCP, but works with SSH tunnels

* Adds associated command line and rake task options

* Reruns once locally if any files are missing from disconnects

# 1.1.0

* Redo exit codes for arguments, signal hanlding, and exceptions
* Update rake task to handle use given order flag

# 1.0.0

* Stop supporting Ruby < v2.0

# 0.5.2

* Added additional exit codes to multirspec runner. 0 for success, 1 for
  failures, 2 for some workers died, 4 for some files didn't get run.

# 0.5.1

* Added flag '--use-given-order' that respects the order files are passed on the
  commandline. The default is still to sort by file size.
* Removed some unneeded RSpec 2 configuration.

# 0.5.0

* Added support for RSpec 3. Did not maintain compatability with RSpec 2.

# 0.4.2

* Change to Rake.application.last_description as last_comment is deprecated

# 0.4.1

* Use RSpec::Core::Runner.disable_autorun! to avoid error messages when workers
  finish

# 0.4.0

* Change TEST_ENV_NUMBER values to match parallel_tests (#10)
* Allow for environment-variable-based defaults for worker count and first-is-1
  in `multirspec` (like #10, improving compatibility with parallel_tests)

# 0.3.0

* Require a filename for the `--log-failing-files` option (#9)
* Exit status for `multirspec` is only 0 on success and is always non-zero when
  there is any sort of failure (#8)

# 0.2.3

* Add `--log-failing-files` option (#7)

# 0.2.2

* Run specs in order of decreasing file size to increase worker utilization

# 0.2.1

* Include filename and line number in realtime output (#5)

# 0.2.0

* Terminate workers when the coordinator process is interrupted with SIGINT (^C)
  or SIGTERM (`kill` with no args) (#3)
* Add per-example timeout option. When using `multirspec` or the rake tasks, it
  defaults to 15 seconds (#2)
* **Breaking change**: per-file timeout is now disabled by default (#2)
* **Breaking change**: changed arguments for `Coordinator`'s constructor
* Correct signal used when requesting that stalled processes quit (use TERM
  instead of QUIT)
* Stop idle processes once there's no work, instead of waiting and stopping
  everything when the suite is complete

# 0.1.0

* Initial version


lib/rspec/multiprocess_runner/command_line_options.rb
require 'rspec/multiprocess_runner'
require 'optparse'
require 'pathname'

module RSpec::MultiprocessRunner
  class CommandLineOptions
    attr_accessor :worker_count, :file_timeout_seconds, :example_timeout_seconds,
      :rspec_options, :explicit_files_or_directories, :pattern, :log_failing_files,
      :first_is_1, :use_given_order, :port, :head_node, :hostname, :max_nodes,
      :run_identifier, :summary_filename

    DEFAULT_WORKER_COUNT = 3

    def initialize
      self.worker_count = default_worker_count
      self.file_timeout_seconds = nil
      self.example_timeout_seconds = 15
      self.pattern = "**/*_spec.rb"
      self.log_failing_files = nil
      self.rspec_options = []
      self.first_is_1 = default_first_is_1
      self.use_given_order = false
      self.port = 2222
      self.hostname = "localhost"
      self.head_node = true
      self.max_nodes = 5
      self.summary_filename = ENV['SUMMARY_FILE']
    end

    def parse(command_line_args, error_stream=$stderr)
      args = command_line_args.dup
      parser = build_parser

      begin
        parser.parse!(args)
      rescue OptionParser::ParseError => e
        error_stream.puts e.to_s
        error_stream.puts parser
        return nil
      end

      if help_requested?
        error_stream.puts parser
        return nil
      end

      extract_files_and_rspec_options(args)
      self
    end

    def files_to_run
      self.explicit_files_or_directories = %w(.) unless explicit_files_or_directories
      relative_root = Pathname.new('.')
      explicit_files_or_directories.map { |path| Pathname.new(path) }.flat_map do |path|
        if path.file?
          path.to_s
        else
          Dir[path.join(pattern).relative_path_from(relative_root)]
        end
      end
    end

    private

    def default_worker_count
      from_env = ENV['MULTIRSPEC_WORKER_COUNT'] || ENV['PARALLEL_TEST_PROCESSORS']
      if from_env
        from_env.to_i
      else
        DEFAULT_WORKER_COUNT
      end
    end

    def default_first_is_1
      from_env = ENV['MULTIRSPEC_FIRST_IS_1'] || ENV['PARALLEL_TEST_FIRST_IS_1']
      %w(true 1).include?(from_env)
    end

    def help_requested!
      @help_requested = true
    end

    def help_requested?
      @help_requested
    end

    def print_default(default_value)
      if default_value
        "default: #{default_value}"
      else
        "none by default"
      end
    end

    def build_parser
      OptionParser.new do |parser|
        parser.banner = "#{File.basename $0} [options] [files or directories] [-- rspec options]"

        parser.on("-w", "--worker-count COUNT", Integer, "Number of workers to run (#{print_default worker_count})") do |n|
          self.worker_count = n
        end

        parser.on("-t", "--file-timeout SECONDS", Float, "Maximum time to allow any single file to run (#{print_default file_timeout_seconds})") do |s|
          self.file_timeout_seconds = s
        end

        parser.on("-T", "--example-timeout SECONDS", Float, "Maximum time to allow any single example to run (#{print_default example_timeout_seconds})") do |s|
          self.example_timeout_seconds = s
        end

        parser.on("-P", "--pattern PATTERN", "A glob to use to select files to run (#{print_default pattern})") do |pattern|
          self.pattern = pattern
        end

        parser.on("--log-failing-files FILENAME", "Filename to log failing files to") do |filename|
          self.log_failing_files = filename
        end

        parser.on("--first-is-1", "Use \"1\" for the first worker's TEST_ENV_NUMBER (instead of \"\")#{" (enabled in environment)" if first_is_1}") do
          self.first_is_1 = true
        end

        parser.on("-O", "--use-given-order", "Use the order that the files are given as arguments") do
          self.use_given_order = true
        end

        parser.on("-p", "--port PORT", Integer, "Communicate using port (#{print_default port})") do |port|
          self.port = port
        end

        parser.on("-H", "--hostname HOSTNAME", "Hostname of the head node (#{print_default hostname})") do |hostname|
          self.hostname = hostname
        end

        parser.on("-n", "--node", "This node is controlled by a head node") do
          self.head_node = false
        end

        parser.on("-m", "--max-nodes MAX_NODES", Integer, "Maximum number of nodes (excluding master) permitted (#{print_default max_nodes})") do |max_nodes|
          self.max_nodes = max_nodes
        end

        parser.on("-r", "--run-identifier STRING", "A unique string used by nodes to confirm identity (e.g. a git commit hash)") do |string|
          self.run_identifier = string
        end

        parser.on("-f", "--summary-filename STRING", "Save summary to a file") do |string|
          self.summary_filename = string
        end

        parser.on_tail("-h", "--help", "Prints this help") do
          help_requested!
        end
      end
    end

    def extract_files_and_rspec_options(remaining_args)
      files = []
      rspec_options = []
      target = files
      remaining_args.each do |arg|
        if arg[0] == '-'
          target = rspec_options
        end
        target << arg
      end

      self.explicit_files_or_directories = files unless files.empty?
      self.rspec_options = rspec_options unless rspec_options.empty?
    end
  end
end
